Introductory_modules/Introduction_to_Ray.ipynb

```diff
@@ -69,26 +68,371 @@
+    "        # update times invoked\n",
+    "        self.invokers[name] += 1\n",
+    "        # return the state of that invoker\n",
+    "        return self.invokers[name]\n",
+    "        \n",
+    "    def get_invoker_state(self, name):\n",
+    "        # return the state of the named invoker\n",
+    "        return self.invokers[name]\n",
+    "    \n",
+    "    def get_all_invoker_state(self):\n",
+    "        # reeturn the state of all invokers\n",
+    "        return self.invokers"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create an instance of our Actor \n",
+    "worker_invoker = MethodStateCounter.remote()\n",
+    "worker_invoker"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Iterate and call the `invoke()` method by random callers and keep track of who called it."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for _ in range(10):\n",
+    "    name = random.choice(CALLERS)\n",
+    "    worker_invoker.invoke.remote(name)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Invoke a random caller and fetch the value or invocations of a random caller."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "for _ in range(5): \n",
+    "    random_name_invoker = random.choice(CALLERS)\n",
+    "    times_invoked = ray.get(worker_invoker.invoke.remote(random_name_invoker))\n",
+    "    print(f\"Named caller: {random_name_invoker} called {times_invoked}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Fetch the count of all callers."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(ray.get(worker_invoker.get_all_invoker_state.remote()))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note: We didn't have to reason about where and how the actors are scheduled. We did not have to worry about the socket connection or IP addresses where these actors reside. All of that is abstracted away from us. All we did is write Python code, using Ray core APIs, convert our classes into distributed stateful services."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 4. Housing Prices with `sklearn`\n",
     "***\n",
-    "Let's take a look at how we can apply Ray Core's flexible and simple API to scale a bare bones version of a common ML task: cross-validation.\n",
+    "Now that we've warmed-up with some isolated examples using tasks and actors, let's take a look at how we can apply Ray Core's flexible and simple API to scale a bare bones version of a common ML task: cross-validation.\n",
```

Comment:
Now that we've warmed-up with some isolated examples using tasks, objects, and actors, let's ....

Reply:
Done!